File: wrc-server.c
#pragma comment(lib, "ws2_32.lib")

#include <assert.h>
#include <stdio.h>
#include <winsock2.h>
#include <windows.h>

#include "wrc-protocol.h"
#include "common.h"

#define logf(fmt,...) printf(fmt "\n", ##__VA_ARGS__)
#define errorf(fmt,...) printf("Error: " fmt "\n", ##__VA_ARGS__)

typedef struct {
  SOCKET sock;
  unsigned char buffer[100];
  unsigned data_len;
} Client;

Client Client_initInvalid()
{
  Client client = {INVALID_SOCKET};
  return client;
}

void Client_setSock(Client *client, SOCKET sock)
{
  client->sock = sock;
  client->data_len = 0;
}



// assumption: len > 0
// returns: length of command on success
//          0 if a partial command was received,
//          len + 1 on error (logs errors)
static unsigned ProcessCommand(unsigned char *cmd, unsigned len)
{
  if (cmd[0] == MOUSE_MOVE) {
    if (len < 9)
      return 0; // need more data
    LONG x = BytesToI32BigEndian(cmd+1);
    LONG y = BytesToI32BigEndian(cmd+5);
    logf("WARNING: mouse move %ld x %ld not implemented", x, y);
    return 9;
  }
  /*
  if (cmd[0] == 'a') {
    logf("got 'a'");
    return 1;
  }
  if (cmd[0] == '\n') {
    logf("got newline");
    return 1;
  }
  */
  errorf("unknown comand 0x%02x", cmd[0]);
  return len + 1; // error
}

// returns: len + 1 on error and logs errors
static unsigned ProcessClientData(unsigned char *data, unsigned len)
{
  unsigned offset = 0;
  for (;;) {
    unsigned remaining = len - offset;
    unsigned result = ProcessCommand(data + offset, remaining);
    if (result == remaining + 1)
      return len + 1; // fail
    if (result == 0)
      return offset;
    offset += result;
    if (offset == len)
      return len;
  }
}

static void HandleClientSock(Client *client)
{
  int len = recv(client->sock, (char*)client->buffer + client->data_len, sizeof(client->buffer) - client->data_len, 0);
  if (len <= 0) {
    if (len == 0) {
      logf("client closed connection");
    } else {
      int error = GetLastError();
      if (error == WSAECONNRESET) {
        logf("client closed connection");
      } else {
        logf("recv function failed with %d", GetLastError());
      }
    }
    closesocket(client->sock);
    client->sock = INVALID_SOCKET;
    return;
  }
  //logf("[DEBUG] got %d bytes", len);
  unsigned total = client->data_len + len;
  unsigned processed = ProcessClientData(client->buffer, total);
  if (processed == total + 1) {
    // error already logged
    closesocket(client->sock);
    client->sock = INVALID_SOCKET;
    return;
  }
  unsigned leftover = total - processed;
  if (leftover > 0) {
    memmove(client->buffer, client->buffer + processed, leftover);
  }
  client->data_len = leftover;
}

static passfail HandleListenSock(SOCKET listen_sock, Client *client)
{
  sockaddr_in from;
  int fromlen = sizeof(from);
  SOCKET new_sock = accept(listen_sock, (sockaddr*)&from, &fromlen);
  if (new_sock == INVALID_SOCKET) {
    errorf("accept function failed with %lu", GetLastError());
    return fail;
  }
  logf("accepted connection from 0x%08lx port %u",
       ntohl(from.sin_addr.s_addr), ntohs(from.sin_port));
  if (client->sock == INVALID_SOCKET) {
    Client_setSock(client, new_sock);
  } else {
    logf("refusing new client (already have client)");
    shutdown(new_sock, SD_BOTH);
    closesocket(new_sock);
  }
  return pass;
}

static passfail ConfigureListenSock(SOCKET s, sockaddr_in* addr)
{
  // TODO: do I need to set reuseaddr socket option?
  if (-1 == bind(s, (sockaddr*)addr, sizeof(*addr))) {
    errorf("bind to address 0x%08lx port %u failed with %lu", ntohl(addr->sin_addr.s_addr),
           ntohs(addr->sin_port), GetLastError());
    return fail;
  }
  if (-1 == listen(s, 0)) {
    errorf("listen function failed with %lu",GetLastError());
    return fail;
  }
  if (pass != SetNonBlocking(s)) {
    errorf("ioctlsocket function to set non-blocking failed with %lu", GetLastError());
    return fail;
  }
  return pass;
}

static SOCKET ListenPortNetworkOrder(u_short port_network_order)
{
  sockaddr_in addr;
  addr.sin_family = AF_INET;
  addr.sin_addr.s_addr = 0;
  addr.sin_port = port_network_order;

  SOCKET s = socket(addr.sin_family, SOCK_STREAM, IPPROTO_TCP);
  if (s == INVALID_SOCKET) {
    errorf("socket function failed with %lu", GetLastError());
    return INVALID_SOCKET;
  }
  if (pass != ConfigureListenSock(s, &addr)) {
    closesocket(s);
    return INVALID_SOCKET;
  }
  return s;
}

void ServeLoop(SOCKET listen_sock)
{
  Client client = Client_initInvalid();
  
  while (true) {
    FD_SET read_set;
    FD_ZERO(&read_set);
    FD_SET(listen_sock, &read_set);
    if (client.sock != INVALID_SOCKET) {
      FD_SET(client.sock, &read_set);
    }
    int popped = select(0, &read_set, NULL, NULL, NULL);
    if (popped == SOCKET_ERROR) {
      errorf("select function failed with %lu", GetLastError());
      return;
    }
    if (popped == 0) {
      errorf("select returned 0?");
      return;
    }

    int handled = 0;
    if (client.sock != INVALID_SOCKET && FD_ISSET(client.sock, &read_set)) {
      HandleClientSock(&client);
      handled++;
    }
    if (handled < popped && FD_ISSET(listen_sock, &read_set)) {
      if (pass != HandleListenSock(listen_sock, &client)) {
        return; // error already logged
      }
    }
  }
}

int main(int argc, char *argv[])
{
  {
    int error = CallWSAStartup();
    if (error != 0) {
      errorf("WSAStartup failed with %lu", GetLastError());
      return 1;
    }
  }

  u_short port = 1234;
  SOCKET listen_sock = ListenPortNetworkOrder(htons(port));
  if (listen_sock == INVALID_SOCKET)
    return 1; // error already logged
  logf("listening on port %u", port);
  ServeLoop(listen_sock);
  return 1;
}

File: wrc-server.zig
const std = @import("std");

pub const UNICODE = true;

const WINAPI = std.os.windows.WINAPI;

const win32 = struct {
    usingnamespace @import("win32").zig;
    usingnamespace @import("win32").foundation;
    usingnamespace @import("win32").system.diagnostics.debug;
    usingnamespace @import("win32").system.system_services;
    usingnamespace @import("win32").ui.windows_and_messaging;
    usingnamespace @import("win32").networking.win_sock;
    usingnamespace @import("win32").ui.display_devices;
};

const proto = @import("wrc-proto.zig");
const common = @import("common.zig");

// NOTE: INPUT does not generate correctly yet because unions are implemented in zigwin32 yet
const win_input = @import("win32").ui.keyboard_and_mouse_input;
const INPUT = extern struct {
    type: win_input.INPUT_TYPE,
    data: extern union {
        mi: extern struct {
            dw: i32,
            dy: i32,
            mouseData: u32,
            dwFlags: u32,
            time: u32,
            dwExtraInfo: usize,
        },
        ki: extern struct {
            wVk: u16,
            wScan: u16,
            dwFlags: u32,
            time: u32,
            dwExtraInfo: usize,
        },
    },
};

const global = struct {
    pub var screen_size: win32.POINT = undefined;
};

const Client = struct {
    sock: win32.SOCKET,
    leftover: [proto.max_msg_data_len-1]u8,
    leftover_len: usize,
    mouse_point: ?win32.POINT,

    pub fn initInvalid() Client {
        return .{
            .sock = win32.INVALID_SOCKET,
            .leftover = undefined,
            .leftover_len = undefined,
            .mouse_point = undefined,
        };
    }

    pub fn setSock(self: *Client, sock: win32.SOCKET) void {
        self.sock = sock;
        self.leftover_len = 0;
        self.mouse_point = null;
    }
};

fn processMessage(client: *Client, msg: proto.ClientToServerMsg, data: []const u8) !void {
    switch (msg) {
    .mouse_move => {
        std.debug.assert(data.len == 8);
        const x = std.mem.readIntBig(i32, data[0..4]);
        const y = std.mem.readIntBig(i32, data[4..8]);
        // TODO: need a way to disable this, possibly with a key press
        if (0 == win32.SetCursorPos(x, y)) {
            std.log.err("SetCursorPos {} {} failed with {}", .{x, y, win32.GetLastError()});
        }
        client.mouse_point = .{ .x = x, .y = y };
    },
    .mouse_button => {
        std.debug.assert(data.len == 2);
        const button: enum {left,right} = switch (data[0]) {
            proto.mouse_button_left => .left,
            proto.mouse_button_right => .right,
            else => |val| {
                std.log.err("unknown mouse button {}", .{val});
                return error.InvalidMessageData;
            },
        };
        const down = switch (data[1]) {
            0 => false,
            1 => true,
            else => |val| {
                std.log.err("expected 0 or 1 for mouse_button down argument but got {}", .{val});
                return error.InvalidMessageData;
            },
        };
        if (client.mouse_point) |mouse_point| {
            const flag = switch (button) {
                .left  => if (down) win_input.MOUSEEVENTF_LEFTDOWN
                          else      win_input.MOUSEEVENTF_LEFTUP,
                .right => if (down) win_input.MOUSEEVENTF_RIGHTDOWN
                          else      win_input.MOUSEEVENTF_RIGHTUP,
            };
            // TODO: this should be const but SendInput arg types is not correct
            var input = INPUT {
                .type = .MOUSE,
                .data = .{ .mi = .{
                    .dw = mouse_point.x,
                    .dy = mouse_point.y,
                    .dwFlags =
                        @enumToInt(flag) |
                        @enumToInt(win_input.MOUSEEVENTF_ABSOLUTE),
                    .mouseData = 0,
                    .time = 0,
                    .dwExtraInfo = 0,
                } },
            };
            if (1 != win_input.SendInput(1, @ptrCast([*]win_input.INPUT, &input), @sizeOf(INPUT))) {
                std.log.err("SendInput failed with {}", .{win32.GetLastError()});
            }
        } else {
            std.log.info("dropping mouse button {} down={}, no mouse position", .{button, down});
        }
    },
    .mouse_wheel => {
        std.debug.assert(data.len == 2);
        const delta = std.mem.readIntBig(i16, data[0..2]);
        if (client.mouse_point) |mouse_point| {
            // TODO: this should be const but SendInput arg types is not correct
            var input = INPUT {
                .type = .MOUSE,
                .data = .{ .mi = .{
                    .dw = mouse_point.x,
                    .dy = mouse_point.y,
                    .dwFlags = @enumToInt(win_input.MOUSEEVENTF_WHEEL),
                    .mouseData = @bitCast(u32, @intCast(i32, delta)),
                    .time = 0,
                    .dwExtraInfo = 0,
                } },
            };
            if (1 != win_input.SendInput(1, @ptrCast([*]win_input.INPUT, &input), @sizeOf(INPUT))) {
                std.log.err("SendInput failed with {}", .{win32.GetLastError()});
            }
        } else {
            std.log.info("dropping mouse wheel {}, no mouse position", .{delta});
        }
    },
    .key => {
        std.debug.assert(data.len == 8);
        const virt_keycode = std.mem.readIntBig(u16, data[0..2]);
        const scan_keycode = std.mem.readIntBig(u16, data[2..4]);
        const flags = std.mem.readIntBig(u32, data[4..8]);
        // TODO: this should be const but SendInput arg types is not correct
        var input = INPUT {
            .type = .KEYBOARD,
            .data = .{ .ki = .{
                .wVk = virt_keycode,
                .wScan = scan_keycode,
                .dwFlags = flags,
                .time = 0,
                .dwExtraInfo = 0,
            } },
        };
        if (1 != win_input.SendInput(1, @ptrCast([*]win_input.INPUT, &input), @sizeOf(INPUT))) {
            std.log.err("SendInput failed with {}", .{win32.GetLastError()});
        }
    },
    }
}

fn processClientData(client: *Client, data: []const u8) ?usize {
    std.debug.assert(data.len > 0);
    var offset: usize = 0;
    while (true) {
        const msg_info = proto.getClientToServerMsgInfo(data[offset]) orelse {
            std.log.err("unknown message id {}", .{data[offset]});
            return null; // fail
        };
        const msg_data_offset = offset + 1;
        if (msg_data_offset + @intCast(usize, msg_info.data_len) > data.len) {
            return offset;
        }
        processMessage(client, msg_info.id, data[msg_data_offset..msg_data_offset + msg_info.data_len]) catch {
            return null; // fail, error already logged
        };
        offset = msg_data_offset + msg_info.data_len;
        if (offset == data.len)
            return data.len;
        std.debug.assert(offset < data.len);
    }
}

fn handleClientSock(client: *Client) void {
    var buffer_storage: [1024]u8 = undefined;
    const buffer: []u8 = &buffer_storage;

    @memcpy(buffer.ptr, &client.leftover, client.leftover_len);

    // TODO: the data type for recv is not correct
    const len = common.tryRecv(client.sock, buffer[client.leftover_len..]) catch |e| {
        switch (e) {
            error.SocketShutdown => std.log.info("client closed connection", .{}),
            error.RecvFailed => std.log.info("recv function failed with {}", .{win32.GetLastError()}),
        }
        if (win32.closesocket(client.sock) != 0) unreachable;
        client.sock = win32.INVALID_SOCKET;
        return;
    };
    //std.log.info("[DEBUG] got {} bytes", .{len});
    const total = client.leftover_len + len;
    const processed = processClientData(client, buffer[0..total]) orelse {
        // error already logged
        if (win32.closesocket(client.sock) != 0) unreachable;
        client.sock = win32.INVALID_SOCKET;
        return;
    };
    client.leftover_len = total - processed;
    std.debug.assert(client.leftover_len < proto.max_msg_data_len);
    @memcpy(&client.leftover, buffer.ptr + processed, client.leftover_len);
}

fn handleListenSock(listen_sock: win32.SOCKET, client: *Client) !void
{
    var from: std.net.Address = undefined;
    var fromlen: i32 = @sizeOf(@TypeOf(from));
    const new_sock = win32.accept(listen_sock, @ptrCast(*win32.SOCKADDR, &from), &fromlen);
    if (new_sock == win32.INVALID_SOCKET) {
        std.log.err("accept function failed with {}", .{win32.GetLastError()});
        return error.AlreadyReported;
    }
    std.log.info("accepted connection from {}", .{from});
    if (client.sock == win32.INVALID_SOCKET) {
        var msg: [8]u8 = undefined;
        std.mem.writeIntBig(i32, msg[0..4], global.screen_size.x);
        std.mem.writeIntBig(i32, msg[4..8], global.screen_size.y);
        common.sendFull(new_sock, &msg) catch {
            std.log.err("failed to send screen size to client, error={}", .{win32.GetLastError()});
            _ = win32.shutdown(new_sock, win32.SD_BOTH);
            if (win32.closesocket(new_sock) != 0) unreachable;
        };
        // TODO: set socket to nonblocking??
        client.setSock(new_sock);
    } else {
        std.log.info("refusing new client (already have client)", .{});
        _ = win32.shutdown(new_sock, win32.SD_BOTH);
        if (win32.closesocket(new_sock) != 0) unreachable;
    }
}

fn FD_ISSET(s: win32.SOCKET, set: *const win32.fd_set) bool {
    for (set.fd_array[0..set.fd_count]) |set_sock| {
        if (set_sock == s)
            return true;
    }
    return false;
}

fn serveLoop(listen_sock: win32.SOCKET) !void {
    var client = Client.initInvalid();

    while (true) {
        var read_set: win32.fd_set = undefined;
        read_set.fd_count = 1;
        read_set.fd_array[0] = listen_sock;
        if (client.sock != win32.INVALID_SOCKET) {
            read_set.fd_count += 1;
            read_set.fd_array[1] = client.sock;
        }
        const popped = win32.select(0, &read_set, null, null, null);
        if (popped == win32.SOCKET_ERROR) {
            std.log.err("select function failed with {}", .{win32.GetLastError()});
            return;
        }
        if (popped == 0) {
            std.log.err("select returned 0?", .{});
            return;
        }

        var handled: u8 = 0;
        if (client.sock != win32.INVALID_SOCKET and FD_ISSET(client.sock, &read_set)) {
            handleClientSock(&client);
            handled += 1;
        }
        if (handled < popped and FD_ISSET(listen_sock, &read_set)) {
            try handleListenSock(listen_sock, &client);
        }
    }
}

pub fn main() !u8 {
    main2() catch |e| switch (e) {
        error.AlreadyReported => return 1,
        else => return e,
    };
    return 0;
}

fn main2() !void {
    global.screen_size = common.getScreenSize();

    // call show cursor, I think this will solve an issue where the cursor goes away

    if (common.wsaStartup()) |e| {
        std.log.err("WSAStartup failed with {}", .{e});
        return error.AlreadyReported;
    }
    const port = 1234;
    const addr_string = "0.0.0.0";
    var addr = std.net.Address.parseIp(addr_string, port) catch |e| {
        std.log.err("invalid ip address '{s}': {}", .{addr_string, e});
        return error.AlreadyReported;
    };
    const s = win32.socket(addr.any.family, win32.SOCK_STREAM, @enumToInt(win32.IPPROTO.TCP));
    if (s == win32.INVALID_SOCKET) {
        std.log.err("socket function failed with {}", .{win32.GetLastError()});
        return error.AlreadyReported;
    }
    // TODO: do I need to set reuseaddr socket option?
    std.os.bind(s, &addr.any, addr.getOsSockLen()) catch |e| {
        std.log.err("bind to {} failed: {}", .{addr, e});
        return error.AlreadyReported;
    };
    std.os.listen(s, 0) catch |e| {
        std.log.err("listen failed: {}", .{e});
        return error.AlreadyReported;
    };
    common.setNonBlocking(s) catch {
        std.log.err("ioctlsocket function to set non-blocking failed with {}", .{win32.GetLastError()});
        return error.AlreadyReported;
    };

    std.log.info("listening for connections at {}", .{addr});
    try serveLoop(s);
}
